qemu/tests/uefishell.py

```diff
@@ -0,0 +1,147 @@
+from virttest import utils_misc
+from virttest import env_process
+
+
+class UEFIShellTest(object):
+    """
+    Provide basic functions for uefishell test. To use UefiShell.iso
+    which is provided by ovmf package. Boot your VM with this as
+    the CD-ROM image and it should boot into the UEFI shell.
+    """
+
+    def __init__(self, test, params, env):
+        """
+        Init the default values for UEFIShell object
+        :param test: Kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment
+        """
+        self.test = test
+        self.params = params
+        self.env = env
+        self.session = None
+
+    def setup(self, under_fs0):
+        """
+        Pre-process for uefishell environment, launch vm from the
+        UefiShell.iso, and setup a serial session for uefishell
+
+        :param under_fs0: most uefi command executed under fs0:\
+        """
+        params = self.params
+        ovmf_path = params["ovmf_path"]
+        uefishell_filename = "UefiShell.iso"
+        uefishell_dst_path = "images/%s" % uefishell_filename
+        params["cdroms"] = "%s %s" % (
+            params["cdroms"], "uefishell")
+        params["cdrom_uefishell"] = uefishell_dst_path
+        uefishell_src_path = utils_misc.get_path(
+            ovmf_path, uefishell_filename)
+        uefishell_dst_path = utils_misc.get_path(
+            data_dir.get_data_dir(), uefishell_dst_path)
+        if not os.path.exists(uefishell_dst_path):
+            cp_command = "cp -f %s %s" % (
+                uefishell_src_path, uefishell_dst_path)
+            process.system(cp_command)
+
+        # Copy a new VAR file for the vm if need, since the
+        # operations inside uefishell will be recorded into
+        # VAR file and may affact later test cases.
+        if params.get("clean_var") == "yes":
+            self.var_copy()
+
+        params["bootindex_uefishell"] = "0"
+        params["start_vm"] = "yes"
+        params["shell_prompt"] = r"(Shell|FS\d:\\.*)>"
+        params["shell_linesep"] = r"\r\n"
+        env_process.process(self.test, params, self.env,
+                            env_process.preprocess_image,
+                            env_process.preprocess_vm)
+
+        vm = self.env.get_vm(params["main_vm"])
+        self.session = vm.wait_for_serial_login()
+        if under_fs0 == "yes":
+            self.send_command("fs0:")
+
+    def var_copy(self):
+        """
+        Copy a new VAR file for a cleaner environment to test
+        """
+        source_var_path = "/usr/share/OVMF/OVMF_VARS.fd"
+        path = storage.get_image_filename_filesytem(
+            self.params, data_dir.get_data_dir())
+        vars_path = "%s.fd" % path
+        if os.path.exists(source_var_path):
+            logging.debug("Use new var file, copied from %s", source_var_path)
+            cmd = "cp -f %s %s" % (source_var_path, vars_path)
+            process.system(cmd)
```

Comment:
If source_var_path doesn't exist, it wont raise error and use the old var file.

Reply:
@YanhuiMa That's nearly impossible, virttest.qemu_vm will need the source file to install and boot up an ovmf guest, if the guest can be boot up, the source file should exist. By that I mean the file is already been checked by qemu_vm. One should execute the test case linearly and without any manual action, otherwise, hardly to assure any of our tests valid.

Reply:
@YanhuiMa And use the old var file might or might not cause the test case FAIL, which also happens when you manually execute the test, I supposed that this automation script does what manual test does.

Reply:
If that's nearly impossible, why not remove if judgement?

Reply:
@YanhuiMa For avoid error, if it really not exist, just use the old one, and the code will not care if the old one will cause FAIL, just fail it when it should fail.

Reply:
No matter manual or auto , I think the basic requirement of running a case is to ensure to use a new and clean var file, if it is not clean, test results are uncertain, and we can't expect the results.